### PR TITLE
[Shodan] Save results to note

### PR DESCRIPTION
**Pull request by annoyingapt**
### Proposed changes

Created config to give the ability to save the results as a note rather than the observable description. I set the default to false so it doesn't mess with existing processes.

### Checklist

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality using different use cases
- [ ] I added/update the relevant documentation (either on github or on notion)
- [ ] Where necessary I refactored code to improve the overall quality

image

### CODE REVIEW

File: internal-enrichment/shodan/src/shodanImport.py [hunk 1]
```diff
@@ -332,27 +338,50 @@ def _generate_stix_vulnerability(self, data):
...
-        self.stix_objects.append(stix_observable)
-        """
-        {
-                    "source_name": "Shodan",
-                    "url": f"www.shodan.io/host/{data['ip_str']}",
-                    "description": "ceci est une description",
-                    # "external_id": str(data['ip_str']),
-                }
-        """
+        if not self.create_note:
```
Comment (Megafredo) [inline internal-enrichment/shodan/src/shodanImport.py]: Here's a small solution for reducing code using dict expansion and ternary:
```python
        # Upsert Observable
        stix_observable = stix2.IPv4Address(
            id=self.stix_entity["id"],
            type="ipv4-addr",
            value=self.stix_entity["value"],
            custom_properties={
                "x_opencti_external_references": [external_reference],
                **({"x_opencti_description": description} if not self.create_note else {}),
                "x_opencti_score": self.score,
                "x_opencti_labels": labels,
                "x_opencti_created_by_ref": self.shodan_identity["standard_id"],
            },
        )
        self.stix_objects.append(stix_observable)
```

File: internal-enrichment/shodan/src/shodanImport.py [hunk 2]
```diff
@@ -332,27 +338,50 @@ def _generate_stix_vulnerability(self, data):
...
+                    "x_opencti_external_references": [external_reference],
+                    "x_opencti_description": description,
+                    "x_opencti_score": self.score,
+                    "x_opencti_labels": labels,
+                    "x_opencti_created_by_ref": self.shodan_identity["standard_id"],
+                },
+            )
+            self.stix_objects.append(stix_observable)
+            """
+            {
```
Comment (Megafredo) [inline internal-enrichment/shodan/src/shodanImport.py]: Can you take this opportunity to delete this comment ? xD

File: internal-enrichment/shodan/src/shodanImport.py [hunk 3]
```diff
@@ -332,27 +338,50 @@ def _generate_stix_vulnerability(self, data):
...
+                value=self.stix_entity["value"],
+                custom_properties={
+                    "x_opencti_external_references": [external_reference],
+                    "x_opencti_score": self.score,
+                    "x_opencti_labels": labels,
+                    "x_opencti_created_by_ref": self.shodan_identity["standard_id"],
+                },
+            )
+            self.stix_objects.append(stix_observable)
+            stix_note = stix2.Note(
```
Comment (Megafredo) [inline internal-enrichment/shodan/src/shodanImport.py:360]: And put the condition for the note directly, small attention can you add the generate_id for the note ?
```Python
# import 
from pycti import Note

# Example
if self.create_note:
  now = datetime.now().strftime("%Y-%m-%dT%H:%M:%SZ")
  stix_note = stix2.Note(
              type="note",
              id=Note.generate_id(now, description),
              abstract="Shodan Results",
              content=description,
              created_by_ref=self.shodan_identity["standard_id"],
              object_refs=self.stix_entity["id"],
          )
          self.stix_objects.append(stix_note)
```

File: internal-enrichment/shodan/src/shodanImport.py (hunk 1 above)
Comment (annoyingapt) [reply internal-enrichment/shodan/src/shodanImport.py]: Good call, I'll make the changes

**Comment by annoyingapt** [comment]
Also I made a note in https://github.com/OpenCTI-Platform/connectors/issues/2605 mentioning that the ShodanAPI doesn't seem to like passing a pattern to it. I won't be able to look into this, but did you want to take this opportunity to remove "Indicator" from the scope in config sample and the compose yml until it it resolved?

**Comment by SamuelHassine** [comment]
@Megafredo @annoyingapt: I would like the behaviour to have result in the note as default = True please. This is way better than overriding the existing description.

**Pull request merged**